pydantic/main.py

```diff
@@ -798,8 +798,14 @@ def __setattr__(self, name: str, value: Any) -> None:
             # TODO - matching error
             raise ValueError(f'"{self.__class__.__name__}" object has no field "{name}"')
         elif self.model_config.get('extra') == 'allow' and name not in self.model_fields:
-            # SAFETY: __pydantic_extra__ is not None when extra = 'allow'
-            self.__pydantic_extra__[name] = value  # type: ignore
+            if self.model_extra and name in self.model_extra:
+                self.__pydantic_extra__[name] = value  # type: ignore
+            else:
+                attr_exists_on_instance: Any = getattr(self, name, None)
```

Comment:
You want to use:
```python
try:
    getattr(self, name)
except AttributeError:
    attr_exists_on_instance = False
else:
    attr_exists_on_instance = True
```
Currently, if the attr exists and is set to anything falsy (e.g., `False`, `0`, `None`, `''`, `[]`, `{}`), the setter will put the new value in extra rather than on the instance.

Reply:
I'll note that you could also remove the intermediate boolean via:
```python
            ...
            else:
                try:
                    getattr(self, name)
                except AttributeError:
                    # attribute does not already exist on instance, so put it in extra
                    self.__pydantic_extra__[name] = value # type: ignore
                else:
                    # attribute _does_ already exist on instance, and was not in extra, so update it
                    _object_setattr(self, name, value)
```